### ISSUE TITLE
Flash of empty Category when clicking uncached product in list

### ISSUE THREAD

**Issue by revlis-x** [DivanteLtd/vue-storefront#4695]
## Current behavior
In 1.12 when clicking a product in category listing which is not cached (e.g. with saveBandwithOverCache: true or products lazy loaded on scroll), data of current category (getCurrentCategory Getter, filters) vanishes before product page has been loaded leading to a brief moment of layout shift and missing content. Depending on page layout this can be quite irritating and definately was not the case in 1.11.

## Expected behavior
Category Content does not vanish when clicking a product in the listing and rendering of product page is complete.

## Steps to reproduce the issue
1. With fresh browser goto https://demo.storefrontcloud.io/women.html
2. Enable offline mode in browser (to make the effect permanently visible).
3. Click a product.

![Screen Shot 2020-07-30 at 13 44 36](https://user-images.githubusercontent.com/2040118/88919164-f068fc80-d26a-11ea-895e-c693d01bc409.png)

## Version of Vue Storefront

- [x] Vue Storefront 1.12.2
- [ ] Vue Storefront Next

## Can you handle fixing this bug by yourself?

- [ ] YES
- [x] NO

**Comment by revlis-x** [comment]
I debugged a little and found, that the behaviour is introduced by https://github.com/DivanteLtd/vue-storefront/commit/ad6e9586a54b0c4b5198985b1dceb2fc4058bacb. The getter is updated much earlier than the previously used root state. 
Any reason for this particular change @gibkigonzo?

**Comment by pkarw** [comment]
@revlis-x  good point. this is a bug that was overlooked during the tests. Can you please propose a fix?

**Comment by gibkigonzo** [comment]
@revlis-x @pkarw I've added this because I wanted to have always newest category in vuex. Take a look on this diagram:
![Untitled Diagram (2)](https://user-images.githubusercontent.com/42383376/89806640-a7049100-db37-11ea-9bb3-5987016dad5b.jpg)

First is before my change and second is after the change. You can see that when we use `rootState.route.params` to resolve current category then `getCurrentCategory` in asyncData is always outdated, because asyncData is triggered before we enter route. Now when `getCurrentCategory` is outdated then other getters are outdated, for example `getAvailableFilters`. And that was the point why I added this change.

Scenario:
1. add red color filter on PLP
2. go to PDP
3. refresh page
4. return to PLP

Before my change in that scenario we will get all products instead of products with red color. 
1. `category-next/loadCategoryProducts` this method is triggered in asyncData
2. [`categoryMappedFilters`](https://github.com/DivanteLtd/vue-storefront/blob/master/core/modules/catalog-next/store/category/actions.ts#L29) is undefined because we don't have cached filters (in 3. step page was refreshed)
3. [`loadCategoryFilters`](https://github.com/DivanteLtd/vue-storefront/blob/master/core/modules/catalog-next/store/category/actions.ts#L32) is triggered to get filters attributes
4. we want to get `searchQuery` so we use [`getters.getCurrentFiltersFrom`](https://github.com/DivanteLtd/vue-storefront/blob/master/core/modules/catalog-next/store/category/actions.ts#L34). Now here is a problem. Because `categoryMappedFilters` is empty so in `getCurrentFiltersFrom` we need to use [`getters.getAvailableFilters`](https://github.com/DivanteLtd/vue-storefront/blob/master/core/modules/catalog-next/store/category/getters.ts#L121) which has outdated category...

I thought that using newest category as soon as possible (in asyncData) will resolve this problem. But as you can see when we use `getCurrentCategory` also to determine how template should look like then some things can be cleared before we want it.
So after all of this I think we have two options:
1. `getCurrentCategory` will be one step behind in asyncData as it was before. 
```
  getCurrentCategory: (state, getters, rootState, rootGetters) => {
    return getters.getCategoryByParams(rootState.route.params)
  },
```
To solve problem in Scenario :point_up: we need to refresh `categoryMappedFilters` in `loadCategoryProducts` and it should be ok
```
    let categoryMappedFilters = getters.getFiltersMap[searchCategory.id]
    if (!categoryMappedFilters && areFiltersInQuery) { // loading all filters only when some filters are currently chosen and category has no available filters yet
      await dispatch('loadCategoryFilters', searchCategory)
      categoryMappedFilters = getters.getFiltersMap[searchCategory.id]
    }
```
2. `getCurrentCategory` will have newest category, but in that case we would need to create separate getters for template
`getCurrentCategory` => `getTemplateCurrentCategory`
`getAvailableFilters` => `getTemplateAvailableFilters`
`getBreadcrumbs` => `getTemplateBreadcrumbs`
names can be different :P

I think that better option is to step back and go to 1. option. That's because we always used `getCurrentCategory` in template and it may be confusing that there will be different getters for template. And if there will be a need, we can always create `getAsyncDataCurrentCategory` (or different name) which will always have newest category and is not bonded with other getters.

**Comment by pkarw** [comment]
@gibkigonzo it's hard for me to get into the details here, so it's more up to your decision. Your call, will be cool I'm sure about it :)
